core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java

```diff
@@ -90,13 +90,34 @@
 public class CalciteMetaImpl extends MetaImpl {
   static final Driver DRIVER = new Driver();
 
+  private final CalciteMetaTableFactory metaTableFactory;
+  private final CalciteMetaColumnFactory metaColumnFactory;
+
+  @Deprecated // to be removed before 2.0
   public CalciteMetaImpl(CalciteConnectionImpl connection) {
+    this(connection, new CalciteMetaTableFactoryImpl(), new CalciteMetaColumnFactoryImpl());
+  }
+
+  public CalciteMetaImpl(CalciteConnectionImpl connection,
+      @Nullable CalciteMetaTableFactory metaTableFactory,
+      @Nullable CalciteMetaColumnFactory metaColumnFactory) {
     super(connection);
     this.connProps
         .setAutoCommit(false)
         .setReadOnly(false)
         .setTransactionIsolation(Connection.TRANSACTION_NONE);
     this.connProps.setDirty(false);
+
+    if (metaTableFactory != null) {
```

Comment:
Anytime I see this kind of `if` block I wish Java had the elvis operator

Reply:
Indeed

Reply:
Au contraire. The solution is to squash out the nulls.

The elvis operator makes it too easy to build horrible chains of optionals.